grades/tasks.py

```diff
@@ -32,54 +31,38 @@
-    final_grade_values = (
-        FinalGrade.objects
-        .filter(
-            course_run__course__program__financial_aid_availability=True,
-            status=FinalGradeStatus.COMPLETE,
-            passed=True,
-        )
-        .annotate(certificate_count=Count('certificate'))
-        .filter(certificate_count=0)
-        .annotate(exam_run_count=Count('course_run__course__exam_runs'))
-        .values(
-            'id',
-            'user',
-            'user__username',
-            'course_run__course',
-            'course_run__edx_course_key',
-            'exam_run_count',
-        )
-    )
-    # Filter the results for FinalGrades that should have a MicromastersCourseCertificate created.
-    # A FinalGrade should have a certificate if:
-    #   1. No exams were scheduled for the course, or
-    #   2. The course has a scheduled exam and the user has a passing ProctoredExamGrade in the course.
-    final_grades_to_certify = filter(
-        lambda final_grade_dict: (
-            final_grade_dict['exam_run_count'] == 0 or
-            ProctoredExamGrade.for_user_course(
-                final_grade_dict['user'],
-                final_grade_dict['course_run__course'])
-            .filter(
-                passed=True
-            ).exists()
-        ),
-        final_grade_values
+    courses = Course.objects.filter(
+        program__live=True,
+        program__financial_aid_availability=True
     )
-    for final_grade_dict in final_grades_to_certify:
-        MicromastersCourseCertificate.objects.create(
-            final_grade_id=final_grade_dict['id']
-        )
-        log.info(
-            'Created MM course certificate for [%s] in course run [%s]',
-            final_grade_dict['user__username'],
-            final_grade_dict['course_run__edx_course_key']
+    for course in courses:
+        if not course.has_frozen_runs():
+            continue
+        users_with_cert = set(
+            MicromastersCourseCertificate.objects.filter(course=course).values_list('user', flat=True)
         )
+        # Find users that passed the course but don't have a certificate yet
+        users_need_cert = list(FinalGrade.objects.filter(
+            course_run__course=course,
+            status=FinalGradeStatus.COMPLETE,
+            passed=True
+        ).exclude(user__in=users_with_cert).values_list('user', flat=True))
+        if course.has_exam:
+            # need also to pass exam
+            users_need_cert = list(ProctoredExamGrade.objects.filter(
+                course=course,
+                passed=True,
+                exam_run__date_grades_available__lte=now_in_utc(),
+                user__in=users_need_cert
+            ).values_list('user', flat=True))
+
+        for user in users_need_cert:
+            MicromastersCourseCertificate.objects.create(
+                user_id=user,
+                course=course
+            )
```

Comment:
This updated logic doesn't account for users that have `FinalGrade`s in multiple course runs in the same course. When that's the case, the `FinalGrade` query above returns 2 identical user id's, and the for loop tries to create a `MicromastersCourseCertificate` for each of them, resulting in the same hash value and an integrity error. I tinkered with the original query in a django shell to see what could be done in a single query, but the result is probably not readable enough. I do like the idea of iterating through courses though, and I came up with the snippet below which limits the number of queries being executed (no separate query for users that have a cert, no `list()`/`set()` usage which executes the queries right away). I added `related_name='user_course_certificates'` to the `MicromastersCourseCertificate.user` definition just for readability. Let me know what you think.

```python
from django.db.models import Q, Count
#...
users_needing_cert_qset = (
  FinalGrade.objects
  .filter(
    course_run__course=course,
    status=FinalGradeStatus.COMPLETE,
    passed=True
  )
  .annotate(
    certificate_count=Count(
      'user__user_certificates',
      filter=Q(user__user_course_certificates__course=course)
    )
  )
  .filter(certificate_count=0)
  .values_list('user', flat=True)
)
if course.has_exam:
  users_needing_cert_qset = ProctoredExamGrade.objects.filter(
    user__in=users_needing_cert_qset
    #...
  )
for user in users_needing_cert_qset:
  #...
```